src/vm/clrtocomcall.cpp

```diff
@@ -629,6 +631,342 @@ UINT32 CLRToCOMEventCallWorker(ComPlusMethodFrame* pFrame, ComPlusCallMethodDesc
     return 0;
 }
 
+CallsiteDetails CreateCallsiteDetails(_In_ FramedMethodFrame *pFrame)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+        PRECONDITION(CheckPointer(pFrame));
+    }
+    CONTRACTL_END;
+
+    MethodDesc *pMD = pFrame->GetFunction();
+    _ASSERTE(!pMD->ContainsGenericVariables() && pMD->IsRuntimeMethodHandle());
+
+    const BOOL fIsDelegate = pMD->GetMethodTable()->IsDelegate();
+    _ASSERTE(!fIsDelegate && pMD->IsRuntimeMethodHandle());
+
+    MethodDesc *pDelegateMD = nullptr;
+    INT32 callsiteFlags = CallsiteDetails::None;
+    if (fIsDelegate)
+    {
+        // Gather details on the delegate itself
+        DelegateEEClass* delegateCls = (DelegateEEClass*)pMD->GetMethodTable()->GetClass();
+        _ASSERTE(pFrame->GetThis()->GetMethodTable()->IsDelegate());
+
+        if (pMD == delegateCls->m_pBeginInvokeMethod.GetValue())
+        {
+            callsiteFlags |= CallsiteDetails::BeginInvoke;
+        }
+        else
+        {
+            _ASSERTE(pMD == delegateCls->m_pEndInvokeMethod.GetValue());
+            callsiteFlags |= CallsiteDetails::EndInvoke;
+        }
+
+        pDelegateMD = pMD;
+
+        // Get at the underlying method desc for this frame
+        pMD = COMDelegate::GetMethodDesc(pFrame->GetThis());
+        _ASSERTE(pDelegateMD != nullptr
+            && pMD != nullptr
+            && !pMD->ContainsGenericVariables()
+            && pMD->IsRuntimeMethodHandle());
+    }
+
+    if (pMD->IsCtor())
+        callsiteFlags |= CallsiteDetails::Ctor;
+
+    Signature signature;
+    Module *pModule;
+    SigTypeContext typeContext;
+
+    if (fIsDelegate)
+    {
+        _ASSERTE(pDelegateMD != nullptr);
+        signature = pDelegateMD->GetSignature();
+        pModule = pDelegateMD->GetModule();
+
+        // If the delegate is generic, pDelegateMD may not represent the exact instantiation so we recover it from 'this'.
+        SigTypeContext::InitTypeContext(pFrame->GetThis()->GetMethodTable()->GetInstantiation(), Instantiation{}, &typeContext);
+    }
+    else if (pMD->IsVarArg())
+    {
+        VASigCookie *pVACookie = pFrame->GetVASigCookie();
+        signature = pVACookie->signature;
+        pModule = pVACookie->pModule;
+        SigTypeContext::InitTypeContext(&typeContext);
+    }
+    else
+    {
+        // COM doesn't support generics so the type is obvious
+        TypeHandle actualType = TypeHandle{ pMD->GetMethodTable() };
+
+        signature = pMD->GetSignature();
+        pModule = pMD->GetModule();
+        SigTypeContext::InitTypeContext(pMD, actualType, &typeContext);
+    }
+
+    _ASSERTE(!signature.IsEmpty() && pModule != nullptr);
+
+    // Create details
+    return CallsiteDetails{ { signature, pModule, &typeContext }, pFrame, pMD, fIsDelegate };
+}
+
+struct AutoHENUMInternal
```

Comment:
We use `HENUMInternalHolder` everywhere else. Why not use it here?

Reply:
Yeah.. about those holders. They are overly complicated and make debugging really annoying due to unnecessary inheritance. The entire holder architecture is over-engineered for what it does. I will use that here, if that is preferred though.

Reply:
> They are overly complicated

I agree with that. Folks building these holders back in the day tried to avoid code duplication between different holder flavors and ended up with this complicated hierarchy as result...